sciris/sc_parallel.py

```diff
@@ -0,0 +1,260 @@
+            time.sleep(interval*2*np.random.rand()) # Sleeps for an average of refresh seconds, but do it randomly so you don't get locking
+        else: 
+            toohigh = False 
+            if verbose: print(label+'CPU load fine (%0.2f/%0.2f), starting process %s after %i tries' % (currentload, maxload, index, count))
+    return None
+
+
+
+class TaskArgs(ut.prettyobj):
+        ''' A class to hold the arguments for the task -- must match both parallelize() and parallel_task() '''
+        def __init__(self, func, index, iterval, itervar, args, kwargs, maxload):
+            self.func = func
+            self.index = index
+            self.iterval = iterval
+            self.itervar = itervar
+            self.args = args
+            self.kwargs = kwargs
+            self.maxload = maxload
+            return None
+
+
+
+def parallelize(func=None, iterarg=None, itervar=None, args=None, kwargs=None, ncpus=None, maxload=None):
+    '''
+    Shortcut for parallelizing a function. iterarg can be an iterable or an integer;
+    if the latter, it will run the function that number of times.
+    
+    The function can either handle a fixed number of CPUs or allocate dynamically based
+    on load. If ncpus is None and maxload is None, then it will use the number of CPUs
+    returned by multiprocessing; if ncpus is not None, it will use the specified number;
+    if maxload is not None, it will allocate dynamically.
+    
+    Example 1:
+        def f(x):
+            return x*x
+        
+        results = parallelize(func=f, iterarg=[1,2,3], ncpus=3)
+    
+    Example 2:
+        import pylab as pl
+        def myfunc(x, y, i):
+            xy = [x+i*pl.rand(1000), y+i*pl.randn(1000)]
+            return xy
+        n = 5
+        xy = parallelize(myfunc, kwargs={'x':3, 'y':8}, iterarg=n, itervar='i', maxload=0.8)
+        for i in range(n):
+            pl.scatter(xy[i][0], xy[i][1])
+    
+    Version: 2018nov01
+    '''
+    # Handle maxload
+    if ncpus is None and maxload is None:
+        ncpus = mp.cpu_count()
+    
+    # Handle iterarg
+    if not(ut.isiterable(iterarg)):
+        try:
+            iterarg = np.arange(iterarg)
+        except Exception as E:
+            errormsg = 'Could not understand iterarg "%s": not iterable and not an integer: %s' % (iterarg, str(E))
+            raise Exception(errormsg)
+    
+    # Construct argument list
+    argslist = []
+    niters = len(iterarg)
+    for index,iterval in enumerate(iterarg):
+        taskargs = TaskArgs(func, index, iterval, itervar, args, kwargs, maxload)
+        argslist.append(taskargs)
+        
+    # Decide how to run -- not load-balanced, use map
+    if ncpus:
+        multipool = mp.Pool(processes=ncpus)
+        outputlist = multipool.map(parallel_task, argslist)
+        return outputlist
+    
+    # It is load-balanced, use Process/Queue
```

Comment:
It's not entirely clear to me how/why `ncpus` controls the load balancing, they sounded like orthogonal options?

Reply:
@RomeshA That's correct, they are orthogonal...I'm not sure what the question is?